File: src/main/python/net/coderextreme/data/ThreeDTexture.py
print('<!--')
import x3d
print('-->')
X3D0 = x3d.X3D()
X3D0.profile = "Immersive"
X3D0.version = "4.1"
head1 = x3d.head()
meta2 = x3d.meta()
meta2.name = "title"
meta2.content = "ThreeDTexture.x3d"

head1.children.append(meta2)
meta3 = x3d.meta()
meta3.name = "creator"
meta3.content = "John Carlson"

head1.children.append(meta3)
meta4 = x3d.meta()
meta4.name = "generator"
meta4.content = "x3d-tidy V3.0.2, https://www.npmjs.com/package/x3d-tidy"

head1.children.append(meta4)
meta5 = x3d.meta()
meta5.name = "generator"
meta5.content = "manual"

head1.children.append(meta5)

X3D0.head = head1
Scene6 = x3d.Scene()
ProtoDeclare7 = x3d.ProtoDeclare()
ProtoDeclare7.name = "sphereproto"
ProtoInterface8 = x3d.ProtoInterface()
field9 = x3d.field()
field9.accessType = "inputOutput"
field9.type = "SFVec3f"
field9.name = "xtranslation"

ProtoInterface8.field.append(field9)

ProtoDeclare7.ProtoInterface = ProtoInterface8
ProtoBody10 = x3d.ProtoBody()
Transform11 = x3d.Transform()
Shape12 = x3d.Shape()
Appearance13 = x3d.Appearance()
Material14 = x3d.Material()
Material14.diffuseColor = [1,1,1]

Appearance13.material = Material14

Shape12.appearance = Appearance13
Sphere15 = x3d.Sphere()

Shape12.geometry = Sphere15

Transform11.children.append(Shape12)
IS16 = x3d.IS()
connect17 = x3d.connect()
connect17.nodeField = "translation"
connect17.protoField = "xtranslation"

IS16.connect.append(connect17)

Transform11.IS = IS16

ProtoBody10.children.append(Transform11)

ProtoDeclare7.ProtoBody = ProtoBody10

Scene6.children.append(ProtoDeclare7)
ProtoDeclare18 = x3d.ProtoDeclare()
ProtoDeclare18.name = "three"
ProtoInterface19 = x3d.ProtoInterface()
field20 = x3d.field()
field20.accessType = "inputOutput"
field20.type = "SFVec3f"
field20.name = "ytranslation"

ProtoInterface19.field.append(field20)

ProtoDeclare18.ProtoInterface = ProtoInterface19
ProtoBody21 = x3d.ProtoBody()
Transform22 = x3d.Transform()
ProtoInstance23 = x3d.ProtoInstance()
ProtoInstance23.name = "sphereproto"

Transform22.children.append(ProtoInstance23)
ProtoInstance24 = x3d.ProtoInstance()
ProtoInstance24.name = "sphereproto"
fieldValue25 = x3d.fieldValue()
fieldValue25.name = "xtranslation"
fieldValue25.value = "2 0 0"

ProtoInstance24.fieldValue.append(fieldValue25)

Transform22.children.append(ProtoInstance24)
ProtoInstance26 = x3d.ProtoInstance()
ProtoInstance26.name = "sphereproto"
fieldValue27 = x3d.fieldValue()
fieldValue27.name = "xtranslation"
fieldValue27.value = "-2 0 0"

ProtoInstance26.fieldValue.append(fieldValue27)

Transform22.children.append(ProtoInstance26)
IS28 = x3d.IS()
connect29 = x3d.connect()
connect29.nodeField = "translation"
connect29.protoField = "ytranslation"

IS28.connect.append(connect29)

Transform22.IS = IS28

ProtoBody21.children.append(Transform22)

ProtoDeclare18.ProtoBody = ProtoBody21

Scene6.children.append(ProtoDeclare18)
ProtoDeclare30 = x3d.ProtoDeclare()
ProtoDeclare30.name = "nine"
ProtoInterface31 = x3d.ProtoInterface()
field32 = x3d.field()
field32.accessType = "inputOutput"
field32.type = "SFVec3f"
field32.name = "ztranslation"

ProtoInterface31.field.append(field32)

ProtoDeclare30.ProtoInterface = ProtoInterface31
ProtoBody33 = x3d.ProtoBody()
Transform34 = x3d.Transform()
ProtoInstance35 = x3d.ProtoInstance()
ProtoInstance35.name = "three"

Transform34.children.append(ProtoInstance35)
ProtoInstance36 = x3d.ProtoInstance()
ProtoInstance36.name = "three"
fieldValue37 = x3d.fieldValue()
fieldValue37.name = "ytranslation"
fieldValue37.value = "0 2 0"

ProtoInstance36.fieldValue.append(fieldValue37)

Transform34.children.append(ProtoInstance36)
ProtoInstance38 = x3d.ProtoInstance()
ProtoInstance38.name = "three"
fieldValue39 = x3d.fieldValue()
fieldValue39.name = "ytranslation"
fieldValue39.value = "0 -2 0"

ProtoInstance38.fieldValue.append(fieldValue39)

Transform34.children.append(ProtoInstance38)
IS40 = x3d.IS()
connect41 = x3d.connect()
connect41.nodeField = "translation"
connect41.protoField = "ztranslation"

IS40.connect.append(connect41)

Transform34.IS = IS40

ProtoBody33.children.append(Transform34)

ProtoDeclare30.ProtoBody = ProtoBody33

Scene6.children.append(ProtoDeclare30)
ProtoDeclare42 = x3d.ProtoDeclare()
ProtoDeclare42.name = "twentyseven"
ProtoInterface43 = x3d.ProtoInterface()
field44 = x3d.field()
field44.accessType = "inputOutput"
field44.type = "SFVec3f"
field44.name = "ttranslation"

ProtoInterface43.field.append(field44)

ProtoDeclare42.ProtoInterface = ProtoInterface43
ProtoBody45 = x3d.ProtoBody()
Transform46 = x3d.Transform()
ProtoInstance47 = x3d.ProtoInstance()
ProtoInstance47.name = "nine"

Transform46.children.append(ProtoInstance47)
ProtoInstance48 = x3d.ProtoInstance()
ProtoInstance48.name = "nine"
fieldValue49 = x3d.fieldValue()
fieldValue49.name = "ztranslation"
fieldValue49.value = "0 0 2"

ProtoInstance48.fieldValue.append(fieldValue49)

Transform46.children.append(ProtoInstance48)
ProtoInstance50 = x3d.ProtoInstance()
ProtoInstance50.name = "nine"
fieldValue51 = x3d.fieldValue()
fieldValue51.name = "ztranslation"
fieldValue51.value = "0 0 -2"

ProtoInstance50.fieldValue.append(fieldValue51)

Transform46.children.append(ProtoInstance50)
IS52 = x3d.IS()
connect53 = x3d.connect()
connect53.nodeField = "translation"
connect53.protoField = "ttranslation"

IS52.connect.append(connect53)

Transform46.IS = IS52

ProtoBody45.children.append(Transform46)

ProtoDeclare42.ProtoBody = ProtoBody45

Scene6.children.append(ProtoDeclare42)
NavigationInfo54 = x3d.NavigationInfo()
NavigationInfo54.type = ["EXAMINE"]

Scene6.children.append(NavigationInfo54)
Viewpoint55 = x3d.Viewpoint()
Viewpoint55.description = "Rubiks Cube"
Viewpoint55.position = [0,0,12]

Scene6.children.append(Viewpoint55)
ProtoInstance56 = x3d.ProtoInstance()
ProtoInstance56.name = "twentyseven"

Scene6.children.append(ProtoInstance56)

X3D0.Scene = Scene6
f = open("../data/ThreeDTexture.new.python.x3d", mode="w", encoding="utf-8")
f.write(X3D0.XML())
f.close()
f = open("../data/ThreeDTexture.new.python.x3dv", mode="w", encoding="utf-8")
f.write(X3D0.VRML())
f.close()
f = open("../data/ThreeDTexture.new.python.x3dj", mode="w", encoding="utf-8")
f.write(X3D0.JSON())
f.close()

File: src/main/go/net/coderextreme/data/ThreeDTexture.go

package main

import (
	"encoding/xml"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"

	"gox3d/x3d" // Assumes your module is named x3d-go-tester

	// Imports for the CGo-based libxml2 wrapper
	"github.com/lestrrat-go/libxml2"
	"github.com/lestrrat-go/libxml2/xsd"
)

// ... (Helper functions remain the same) ...
func stringPtr(s string) *string { return &s }
func floatPtr(f float32) *float32 { return &f }
func doublePtr(d float64) *float64 { return &d }
func boolPtr(b bool) *bool       { return &b }
func int32Ptr(i int32) *int32    { return &i }

func downloadSchemaIfNotExists(url, filepath string) error {
	// ... (function is correct, no changes) ...
	if _, err := os.Stat(filepath); err == nil {
		fmt.Printf("Schema file '%s' already exists. Skipping download.\n", filepath)
		return nil
	}
	fmt.Printf("Downloading X3D schema from %s to %s...\n", url, filepath)
	resp, err := http.Get(url)
	if err != nil { return fmt.Errorf("failed to download schema: %w", err) }
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK { return fmt.Errorf("bad status: %s", resp.Status) }
	out, err := os.Create(filepath)
	if err != nil { return fmt.Errorf("failed to create schema file: %w", err) }
	defer out.Close()
	_, err = io.Copy(out, resp.Body)
	if err != nil { return fmt.Errorf("failed to save schema: %w", err) }
	fmt.Println("Schema downloaded successfully.")
	return nil
}

// Updated validation function to print detailed errors
func validateXMLWithSchema(xmlData []byte, schemaPath string) error {
	schema, err := xsd.ParseFromFile(schemaPath)
	if err != nil { return fmt.Errorf("failed to parse schema file %s: %w", schemaPath, err) }
	defer schema.Free()
	doc, err := libxml2.Parse(xmlData)
	if err != nil { return fmt.Errorf("failed to parse XML data: %w", err) }
	defer doc.Free()

	validationErr := schema.Validate(doc)
	if validationErr != nil {
		errorLog := ""
		if schemaErrors, ok := validationErr.(xsd.SchemaValidationError); ok {
			for i, e := range schemaErrors.Errors() {
				errorLog += fmt.Sprintf("Error %d: %s\n", i+1, e.Error())
			}
			return fmt.Errorf("schema validation failed with %d errors:\n%s", len(schemaErrors.Errors()), errorLog)
		}
		return fmt.Errorf("schema validation failed: %w", validationErr)
	}
	return nil
}

func main() {
	fmt.Println("--- Building and Testing an X3D Scene in Go ---")

	const schemaURL = "https://www.web3d.org/specifications/x3d-4.0.xsd"
	const schemaFilename = "x3d-4.0.xsd"
	if err := downloadSchemaIfNotExists(schemaURL, schemaFilename); err != nil {
		log.Fatalf("Could not prepare schema file: %v", err)
	}
    sceneRoot := &x3d.X3D{
        Profile: stringPtr("Immersive"),
        Version: stringPtr("4.0"),
        Head: &x3d.Head{
            Metas: []*x3d.Meta{
                &x3d.Meta{
                    Name: stringPtr("title"),
                    Content: stringPtr("ThreeDTexture.x3d"),
            },
            &x3d.Meta{
                Name: stringPtr("creator"),
                Content: stringPtr("John Carlson"),
            },
            &x3d.Meta{
                Name: stringPtr("generator"),
                Content: stringPtr("manual"),
            },
            &x3d.Meta{
                Name: stringPtr("generator"),
                Content: stringPtr("x3d-tidy V2.1.21, https://www.npmjs.com/package/x3d-tidy"),
            },
            },
        },
        Scene: &x3d.Scene{
            Children: []x3d.X3DChildNode{
                &x3d.ProtoDeclare{
                    Name: stringPtr("sphereproto"),
                    ProtoInterface: &x3d.ProtoInterface{
                        Field: []x3d.X3DNode{
                            &x3d.Field{
                                AccessType: stringPtr("inputOutput"),
                                Type: stringPtr("SFVec3f"),
                                Name: stringPtr("xtranslation"),
                        },
                        },
                    },
                    ProtoBody: &x3d.ProtoBody{
                        Children: []x3d.X3DNode{
                            &x3d.Transform{
                                    IS: &x3d.IS{
                                        Connect: []x3d.X3DNode{
                                            &x3d.Connect{
                                                NodeField: stringPtr("translation"),
                                                ProtoField: stringPtr("xtranslation"),
                                        },
                                        },
                                    },
                                Children: []x3d.X3DNode{
                                    &x3d.Shape{
                                        Appearance: &x3d.Appearance{
                                            Material: &x3d.Material{
                                                DiffuseColor: &x3d.SFColor{1.0, 1.0, 1.0},
                                            },
                                        },
                                        Geometry: &x3d.Sphere{
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &x3d.ProtoDeclare{
                    Name: stringPtr("three"),
                    ProtoInterface: &x3d.ProtoInterface{
                        Field: []x3d.X3DNode{
                            &x3d.Field{
                                AccessType: stringPtr("inputOutput"),
                                Type: stringPtr("SFVec3f"),
                                Name: stringPtr("ytranslation"),
                        },
                        },
                    },
                    ProtoBody: &x3d.ProtoBody{
                        Children: []x3d.X3DNode{
                            &x3d.Transform{
                                    IS: &x3d.IS{
                                        Connect: []x3d.X3DNode{
                                            &x3d.Connect{
                                                NodeField: stringPtr("translation"),
                                                ProtoField: stringPtr("ytranslation"),
                                        },
                                        },
                                    },
                                Children: []x3d.X3DNode{
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("sphereproto"),
                                    },
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("sphereproto"),
                                        FieldValue: []x3d.X3DNode{
                                            &x3d.FieldValue{
                                                Name: stringPtr("xtranslation"),
                                                Value: stringPtr("2 0 0"),
                                        },
                                        },
                                    },
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("sphereproto"),
                                        FieldValue: []x3d.X3DNode{
                                            &x3d.FieldValue{
                                                Name: stringPtr("xtranslation"),
                                                Value: stringPtr("-2 0 0"),
                                        },
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &x3d.ProtoDeclare{
                    Name: stringPtr("nine"),
                    ProtoInterface: &x3d.ProtoInterface{
                        Field: []x3d.X3DNode{
                            &x3d.Field{
                                AccessType: stringPtr("inputOutput"),
                                Type: stringPtr("SFVec3f"),
                                Name: stringPtr("ztranslation"),
                        },
                        },
                    },
                    ProtoBody: &x3d.ProtoBody{
                        Children: []x3d.X3DNode{
                            &x3d.Transform{
                                    IS: &x3d.IS{
                                        Connect: []x3d.X3DNode{
                                            &x3d.Connect{
                                                NodeField: stringPtr("translation"),
                                                ProtoField: stringPtr("ztranslation"),
                                        },
                                        },
                                    },
                                Children: []x3d.X3DNode{
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("three"),
                                    },
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("three"),
                                        FieldValue: []x3d.X3DNode{
                                            &x3d.FieldValue{
                                                Name: stringPtr("ytranslation"),
                                                Value: stringPtr("0 2 0"),
                                        },
                                        },
                                    },
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("three"),
                                        FieldValue: []x3d.X3DNode{
                                            &x3d.FieldValue{
                                                Name: stringPtr("ytranslation"),
                                                Value: stringPtr("0 -2 0"),
                                        },
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &x3d.ProtoDeclare{
                    Name: stringPtr("twentyseven"),
                    ProtoInterface: &x3d.ProtoInterface{
                        Field: []x3d.X3DNode{
                            &x3d.Field{
                                AccessType: stringPtr("inputOutput"),
                                Type: stringPtr("SFVec3f"),
                                Name: stringPtr("ttranslation"),
                        },
                        },
                    },
                    ProtoBody: &x3d.ProtoBody{
                        Children: []x3d.X3DNode{
                            &x3d.Transform{
                                    IS: &x3d.IS{
                                        Connect: []x3d.X3DNode{
                                            &x3d.Connect{
                                                NodeField: stringPtr("translation"),
                                                ProtoField: stringPtr("ttranslation"),
                                        },
                                        },
                                    },
                                Children: []x3d.X3DNode{
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("nine"),
                                    },
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("nine"),
                                        FieldValue: []x3d.X3DNode{
                                            &x3d.FieldValue{
                                                Name: stringPtr("ztranslation"),
                                                Value: stringPtr("0 0 2"),
                                        },
                                        },
                                    },
                                    &x3d.ProtoInstance{
                                        Name: stringPtr("nine"),
                                        FieldValue: []x3d.X3DNode{
                                            &x3d.FieldValue{
                                                Name: stringPtr("ztranslation"),
                                                Value: stringPtr("0 0 -2"),
                                        },
                                        },
                                    },
                                },
                            },
                        },
                    },
                },
                &x3d.NavigationInfo{
                    Type: x3d.MFString{"EXAMINE"},
                },
                &x3d.Viewpoint{
                    Description: stringPtr("Rubiks Cube"),
                    Position: &x3d.SFVec3f{0.0, 0.0, 12.0},
                },
                &x3d.ProtoInstance{
                    Name: stringPtr("twentyseven"),
                },
            },
        },
    }


	fmt.Println("\n--- Validating the generated scene (internal logic) ---")
	err := sceneRoot.Scene.Validate()
	if err != nil {
		log.Fatalf("Scene validation failed: %v", err)
	}
	fmt.Println("✅ Internal scene validation successful!")
	fmt.Println("\n--- Serializing scene to XML ---")
	output, err := xml.MarshalIndent(sceneRoot, "", "  ")

	if err != nil {
		log.Fatalf("XML Marshaling failed: %v", err)
	}
	/*
	fmt.Println("\n--- Validating XML against X3D 4.0 Schema (using libxml2) ---")
	err = validateXMLWithSchema(output, schemaFilename)
	if err != nil {
		fmt.Printf("--- Invalid Generated XML ---\n%s\n---------------------------\n", string(output))
		log.Fatalf("Schema validation failed for generated XML: %v", err)
	}
	fmt.Println("✅ XML is valid against the X3D 4.0 schema!")
	*/
	filename := "../data/ThreeDTexture.new.go.x3d"
	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return
	}
	defer file.Close() // Ensure the file is closed when the function exits

	// Write the string content to the file
	header := "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE X3D PUBLIC \"ISO//Web3D//DTD X3D 4.0//EN\" \"https://www.web3d.org/specifications/x3d-4.0.dtd\">\n"
	_, err = file.WriteString(header)
	if err != nil {
		fmt.Printf("Error writing header to file: %v\n", err)
		return
	}

	_, err = file.Write(output)
	if err != nil {
		fmt.Printf("Error writing bytes to file: %v\n", err)
		return
	}

	fmt.Printf("\n✅ Scene successfully written to %s\n", filename)
	fmt.Printf("   You can view the generated XML below:\n\n%s\n", string(output))
}